linux/LinuxProcessList.c

```diff
@@ -1222,49 +1227,65 @@ static inline void LinuxProcessList_scanMemoryInfo(ProcessList* this) {
    fclose(file);
 }
 
-static inline void LinuxProcessList_scanZramInfo(LinuxProcessList* this) {
+static void LinuxProcessList_scanZramInfo(LinuxProcessList* this) {
    unsigned long long int totalZram = 0;
    unsigned long long int usedZramComp = 0;
    unsigned long long int usedZramOrig = 0;
 
-   char mm_stat[34];
-   char disksize[34];
+   DIR* sysBlockDir = opendir("/sys/block");
+   if (!sysBlockDir)
+      goto end;
 
-   unsigned int i = 0;
-   for (;;) {
-      xSnprintf(mm_stat, sizeof(mm_stat), "/sys/block/zram%u/mm_stat", i);
-      xSnprintf(disksize, sizeof(disksize), "/sys/block/zram%u/disksize", i);
-      i++;
-      FILE* disksize_file = fopen(disksize, "r");
-      FILE* mm_stat_file = fopen(mm_stat, "r");
-      if (disksize_file == NULL || mm_stat_file == NULL) {
-         if (disksize_file) {
-            fclose(disksize_file);
-         }
-         if (mm_stat_file) {
-            fclose(mm_stat_file);
-         }
-         break;
+   int sysBlockFd = dirfd(sysBlockDir);
+   const struct dirent* entry;
+   while ((entry = readdir(sysBlockDir)) != NULL) {
+      if (!String_startsWith(entry->d_name, "zram"))
+         continue;
+
+      if (entry->d_name[4] < '0' || entry->d_name[4] > '9')
+         continue;
+
+      int zramFd = openat(sysBlockFd, entry->d_name, O_PATH | O_DIRECTORY);
+      if (zramFd < 0)
+         continue;
+
+      FILE* disksize_file = fopenat(zramFd, "disksize", "r");
+      if (!disksize_file) {
+         close(zramFd);
+         continue;
+      }
+
+      FILE* mm_stat_file = fopenat(zramFd, "mm_stat", "r");
+      if (!mm_stat_file) {
+         fclose(disksize_file);
+         close(zramFd);
+         continue;
       }
+
       unsigned long long int size = 0;
       unsigned long long int orig_data_size = 0;
       unsigned long long int compr_data_size = 0;
 
       if (!fscanf(disksize_file, "%llu\n", &size) ||
           !fscanf(mm_stat_file, "    %llu       %llu", &orig_data_size, &compr_data_size)) {
-         fclose(disksize_file);
          fclose(mm_stat_file);
-         break;
+         fclose(disksize_file);
+         close(zramFd);
+         continue;
```

Comment:
Why not just force `size`, `orig_data_size` and `compr_data_size` all to zero here and fall through?

Or invert this condition and handle the three additions from below in here.

Reply:
dropped the zram changes, no real gain